# Relacionar Bases de Datos

En esta sección buscaremos solucionar el problema de clasificación que existe para el Dataset "Breweries, Beers and Reviews", en específico al relacionar los estilos de cerveza por las clasificaciones de la BJCP, al asignarles el código de identificación único por estilo. 

A continuación relacionaremos la base de datos "beers" (obtenida en https://www.kaggle.com/ehallmar/beers-breweries-and-beer-reviews) con la base de datos "2015_Guidelines" (obtenida en https://www.bjcp.org/stylecenter.php) la primera contiene información sobre 358,873 cervezas clasificadas en 113 estilos, la segunda contiene datos de relación relación sobre 119 estilos de cervezas (Categorias, Porcentajes, Composición, Descripciones, etc.)

In [4]:
import pandas as pd

In [5]:
beers = pd.read_csv('BBR/beers.csv')
style = pd.read_excel('SUP/2015_Guidelines.xlsx')

display(beers.head(3))
display(style.head(3))

id                                  name  brewery_id state country  \
0  202522                        Olde Cogitator        2199    CA      US   
1   82352  Konrads Stout Russian Imperial Stout       18604   NaN      NO   
2  214879                        Scottish Right       44306    IN      US   

                    style availability   abv                   notes retired  
0   English Oatmeal Stout     Rotating   7.3  No notes at this time.       f  
1  Russian Imperial Stout     Rotating  10.4  No notes at this time.       f  
2            Scottish Ale   Year-round   4.0  No notes at this time.       t

#         BJCP Categories                Styles Style Family  \
0  01A  Standard American Beer  American Light Lager   Pale Lager   
1  01B  Standard American Beer        American Lager   Pale Lager   
2  01C  Standard American Beer             Cream Ale     Pale Ale   

              Style History         Origin ABV min ABV max IBUs min IBUs max  \
0    Mass Market Pale Lager  United States     2.8     4.2        8       12   
1    Mass Market Pale Lager  United States     4.2     5.3        8       18   
2  Indigenous American Beer  United States     4.2     5.6        8       20   

   ...                                          Mouthfell  \
0  ...  Very light (sometimes watery) body. Very highl...   
1  ...  Low to medium-low body. Very highly carbonated...   
2  ...  Generally light and crisp, although body can r...   

                                            Comments  \
0  Designed to appeal to as broad a range of the ...   
1  Strong flavors are a fault. Often what non-cra...   
2  Pre-prohibition Cream Ales were slightly stron...   

                                             History  \
0  Coors briefly made a light lager in the early ...   
1  Although German immigrants had brewed traditio...   
2  A sparkling or present-use ale that existed in...   

                          Characteristic Ingredients  \
0  Two- or six-row barley with high percentage (u...   
1  Two- or six-row barley with high percentage (u...   
2  American ingredients most commonly used. A gra...   

                                    Style Comparison  \
0  A lighter-bodied, lower-alcohol, lower calorie...   
1  Stronger, more flavor and body than a Light Am...   
2  Similar to a Standard American Lager, but with...   

                                 Commercial Examples Notes Unnamed: 27  \
0  Bud Light, Coors Light, Keystone Light, Michel...   NaN         NaN   
1  Budweiser, Coors Original, Grain Belt Premium ...   NaN         NaN   
2  Genesee Cream Ale, Liebotschaner Cream Ale, Li...   NaN         NaN   

  Unnamed: 28 Unnamed: 29  
0         NaN         NaN  
1         NaN         NaN  
2         NaN         NaN  

[3 rows x 30 columns]

In [6]:
beers_per_style = beers.groupby('style').size().sort_values(ascending = False).reset_index().rename(columns = {"0":"beers_per_style"})
style_fam = style[['#',"BJCP Categories", "Style Family","Styles"]]

display(beers_per_style.head(3))
print(len(beers_per_style))
      
display(style_fam.head(3))
print(len(style_fam))

style      0
0             American IPA  44719
1  American Pale Ale (APA)  22159
2    American Imperial IPA  18338

112


#         BJCP Categories Style Family                Styles
0  01A  Standard American Beer   Pale Lager  American Light Lager
1  01B  Standard American Beer   Pale Lager        American Lager
2  01C  Standard American Beer     Pale Ale             Cream Ale

118


In [7]:
beers_per_id_style = style_fam.merge(beers_per_style, how='inner', left_on = 'Styles', right_on = 'style')

display(beers_per_id_style)
len(style_fam.merge(beers_per_style, how='inner', left_on = 'Styles', right_on = 'style'))

#                BJCP Categories Style Family                Styles  \
0   01A         Standard American Beer   Pale Lager  American Light Lager   
1   01B         Standard American Beer   Pale Lager        American Lager   
2   07A     Amber Bitter European Beer  Amber Lager          Vienna Lager   
3   09C           Strong European Beer       Porter         Baltic Porter   
4   13C             Brown British Beer       Porter        English Porter   
5   15A                     Irish Beer    Amber Ale         Irish Red Ale   
6   19C  Amber and Brown American Beer    Brown Ale    American Brown Ale   
7   20A      American Porter and Stout       Porter       American Porter   
8   20B      American Porter and Stout        Stout        American Stout   
9   21A                            IPA          IPA          American IPA   
10  22B            Strong American Ale   Strong Ale   American Strong Ale   
11  22C            Strong American Ale   Strong Ale   American Barleywine   
12  23A              European Sour Ale   Wheat Beer       Berliner Weisse   
13  23B              European Sour Ale     Sour Ale      Flanders Red Ale   
14  24B                    Belgian Ale     Pale Ale      Belgian Pale Ale   
15  26B                   Trappist Ale    Amber Ale        Belgian Dubbel   
16  26C                   Trappist Ale   Strong Ale        Belgian Tripel   

                   style      0  
0   American Light Lager   1333  
1         American Lager   3741  
2           Vienna Lager   1567  
3          Baltic Porter   1614  
4         English Porter   2263  
5          Irish Red Ale   2080  
6     American Brown Ale   7008  
7        American Porter  10168  
8         American Stout   9103  
9           American IPA  44719  
10   American Strong Ale   2509  
11   American Barleywine   3264  
12       Berliner Weisse   5036  
13      Flanders Red Ale    774  
14      Belgian Pale Ale   4523  
15        Belgian Dubbel   2376  
16        Belgian Tripel   3783

17

Debido a que los estilos tienen diferentes redacciones o usan sinónimos, requiere una revisión metódica para asignar correctamente los códigos, se puede facilitar el trabajo al relacionar strings, sin embargo una revisión manual, es impresindible para los 118 estilos.

In [8]:
with pd.ExcelWriter("SUP/style_index.xlsx") as writer:
    beers_per_id_style.to_excel(writer, index=False, sheet_name="style")
    beers_per_style.to_excel(writer, index=False, sheet_name="style_beers")
    style_fam.to_excel(writer, index=False, sheet_name="style_fam")

## Asignación de código y comprobación de integridad de datos

Una vez comprobados que los strings esten correctamente relacionados, procedemos a cargar style_index.xlsx para realizar la unión.

In [9]:
style_idx = pd.read_excel('SUP/style_index.xlsx')

Para llevar esto a cabo desarrollamos el dataframe "style_idx" que relaciona las columnas 'style' de la base de datos "beers" y 'Styles' de la base de datos "2015_Guidelines", con la finalidad de asignar el código único de identificación "#".

In [10]:
display(style_idx[['#','Styles','style']].head(3))

#                Styles                 style
0  01A  American Light Lager  American Light Lager
1  01B        American Lager        American Lager
2  07A          Vienna Lager          Vienna Lager

Una vez realizado el merge nuestra base de datos "beers_idx" esta directamente relacionada con "2015_Guidelines", concluimos buscando perdida de información tras realizar esta operación.

In [11]:
beers_idx = beers.merge(style_idx, how = 'inner', left_on = 'style', right_on = 'style').rename(columns = {0:'beers_per_style'})[['id','name','brewery_id','state','country','availability','abv','style','#','beers_per_style']
]

display(beers_idx.head(3))

print('Se conservan datos sobre ' + str(len(beers_idx))[:3] + ',' + str(len(beers_idx))[3:] + ' cervezas de ' + str(len(beers))[:3] + ',' + str(len(beers))[3:] + ' que teníamos inicialmente (' + str(len(beers_idx)/len(beers)*100)[:5] + '%).')

#beers_idx.to_csv('beers_idx.csv')

id                 name  brewery_id state country availability  abv  \
0  108605          Icon Sender       22598    CA      US   Year-round  5.6   
1  255286  Light Of The Ozarks       11203    AR      US     Rotating  4.3   
2  332972               Vibin’       31805    FL      US     Rotating  5.0   

            style    #  beers_per_style  
0  American Lager  01B             3741  
1  American Lager  01B             3741  
2  American Lager  01B             3741

Se conservan datos sobre 105,861 cervezas de 358,873 que teníamos inicialmente (29.49%).
